Updated module with 2.5+ requirements and example

mglantz opened the pull request:
I've updated the documentation with 2.5+ requirements plus a working example of implementing them.

load_balancing_rules requires frontend_port and also requires backend_port to be passed if not the user has taken special action to enable additional features on the azure subscription. I marked both as required as the backend_port error message doesn't really explain that backend_port needs to be defined except for in a special scenario.

Error message as follows
'''
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error creating or updating load balancer myLoadBalancer - Azure Error: SubscriptionNotRegisteredForFeature\nMessage: Subscription asdfasdfasdfasdfasdfasdfasdfasdfasdf is not registered for feature Microsoft.Network/AllowILBAllPortsRule required to carry out the requested operation."}
'''

+label: docsite_pr

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

mglantz committed:
Updated module with 2.5+ requirements and example

I've updated the documentation with 2.5+ requirements plus a working example of implementing them.

load_balancing_rules requires frontend_port and also requires backend_port to be passed if not the user has taken special action to enable additional features on the azure subscription. I marked both as required as the backend_port error message doesn't really explain that backend_port needs to be defined except for in a special scenario.

Error message as follows
'''
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error creating or updating load balancer myLoadBalancer - Azure Error: SubscriptionNotRegisteredForFeature\nMessage: Subscription asdfasdfasdfasdfasdfasdfasdfasdfasdf is not registered for feature Microsoft.Network/AllowILBAllPortsRule required to carry out the requested operation."}
'''

+label: docsite_pr

mglantz committed:
Removed trailing whitespace

yuwzho commented:
shipit

Unfortunately your test together with other tests run out of the subscription's resource limit. Please close this pr and reopen it to trigger another build.

yuwzho commented:
bot_status

yungezz commented on lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py:
```diff
@@ -189,11 +189,13 @@
                     - The port used for internal connections on the endpoint.
                     - Acceptable values are between 0 and 65535.
                     - Note that value 0 enables "Any Port"
+                required: True
```
backend port is not required

mglantz replied:
Hello there, yes, that is true. Please see first post for an argument to set it True. The feature to be activated on the Azure account is not activated by default and the error message is close to garbage for most people. By default it's required, it's not required for people who has activated AllowILBAllPortsRule. IMHO it then makes sense to have it required.

yuwzho replied:
Well, the documentation should align the spec. In the spec the backend_port in inat is required but not in lb rules

https://github.com/ansible/ansible/blob/00884581ffabed74ce9ea3ee9f99a7d34cfffd56/lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py#L461-L463

yungezz commented:
@mglantz  could you pls resolve conflicts?

Fred-sun commented:
@mglantz Thanks for you contribution,  Could you pls resolve conflicts?

mglantz committed:
Removed requirement for backend port

mglantz committed:
Merge branch 'devel' into patch-1

mglantz commented:
Hi all, now back from vacation... :) I removed the requirement for backend port and resolved conflict.

yuwzho commented:
@mglantz It looks ok from my side, but the bot will prevent the merge since there is a merge commit. You may need to rebase your code to devel branch and force push again. Then trigger CI again.

Fred-sun commented:
@yuwzho @yungezz Could you help take a look this PR? Thanks!

yuwzho commented:
@mglantz Seems this is not related to your changes, but also the lb's documentation sanity test failed. From line 352, there need one more indent. Would you mind changing these line also?

yuwzho commented:
from line 310 - line 351, are over indent

https://github.com/ansible/ansible/blob/0eeee742870b66542b14e904d14fb2776f739976/lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py#L310-L351

Fred-sun commented:
@mglantz Thanks for the contribution, Could you help to recheck the PR according by @yuwzho's advice? Thanks !

Fred-sun commented:
@mglantz Thanks for the contribution, Could you help to recheck the PR according by @yuwzho's comment? Thanks !

Fred-sun commented:
@mglantz

Fred-sun commented:
@mglantz Thanks for the contribution, Are you still here? Could you help change according by @yuwzho's comment? Thanks!

mglantz committed:
Fixed indention

It was over indent

mglantz commented:
Hey @Fred-sun, sorry for this taking 1000 years. I've fixed it.

mglantz commented:
@Fred-sun I can't see any errors coming out of Shippable, even though it seems to say that all checks have failed. Let me know. I don't have anything competing for my attention except for this now.

Fred-sun commented:
@mglantz Thanks for you update, As @yuwzho describe, From line 310 - line 351 both need to be indented two space forward. Thanks!

mglantz commented:
@Fred-sun Forward, aah.

Fred-sun commented:
@yuwzho Could you help to review this PR again? The contributor has changed. Thanks!

mglantz committed:
Identeded correctly

I followed how it's been done in https://github.com/mglantz/ansible/blob/patch-1/lib/ansible/modules/cloud/azure/azure_rm_aks.py

mglantz commented:
> @yuwzho Could you help to review this PR again? The contributor has changed. Thanks!

I've changed it so that I follow the same indention as in https://github.com/mglantz/ansible/blob/patch-1/lib/ansible/modules/cloud/azure/azure_rm_aks.py

mglantz commented:
> The test `ansible-test sanity --test validate-modules` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html)] failed with 1 error:
> 
> ```
> lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py:323:7: E311 EXAMPLES is not valid YAML
> ```
> The test `ansible-test sanity --test yamllint` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/yamllint.html)] failed with 1 error:
> 
> ```
> lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py:323:7: error EXAMPLES: syntax error: expected <block end>, but found '-'
> ```
> [click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

Fixing..

yungezz commented on lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py:
```diff
@@ -306,32 +307,55 @@
+      azure_rm_loadbalancer:
+        name: myLoadBalancer
+        resource_group: myResourceGroup
+        location: northeurope
```
location is not necessary

acozine replied:
```suggestion
```

yungezz commented on lib/ansible/modules/cloud/azure/azure_rm_loadbalancer.py:
```diff
@@ -306,32 +307,55 @@
+            protocol: Tcp
+            backend_address_pool: myBackendPool
+            probe: myProbe
+      # Example for <2.5 module
```
i dont think this is necessary to keep old one. docs has version also

Fred-sun commented:
@mglantz Thanks for your contribution, Could you help update according by yungezz's comment? Thanks!

Fred-sun commented:
@mglantz Thanks for the contrition, Are you still here? Could you help give a change according by yungezz and yuwzho's comment? Thanks!

Fred-sun commented:
kindly ping

Fred-sun commented:
@mglantz Thanks for your contribution. Could you help rebase the PR and resolve the comments? Thanks!

Fred-sun commented:
kindly ping

Fred-sun commented:
@mglantz  Are you still here? Could you help rebase the PR and resolve the comments? Thanks!

acozine reviewed (changes requested):
Thanks @mglantz for taking the time to update the Ansible documentation.

I agree with @yungezz that we should remove the example for pre-2.5 - those examples will remain in the 2.5 docs. If you do that, rebase, and ping me back, I'll try to merge this quickly, before other changes go in on the same module.

Fred-sun commented:
kindly ping

Fred-sun commented:
@mglantz Are you still here? Please help resolve the conflicting files and change according by above comments! Thanks!

Fred-sun commented:
@yungezz The PR has no update more than five month, and exists conflicting file, Could we can close this? also can create a new to instead of it when this function need! Thanks!

yungezz commented:
close it for now, due to conflict and no response on comments.  ansible docs has version,  so devel branch is always latest sample. if user need check sample of old version, can check doc  of old version.  pls reopen it if necessary we can continue the discussion.
